From 8ebcddf28e7e887b8d18c9c375a28843a7e2de5d Mon Sep 17 00:00:00 2001
From: Paul Kocialkowski <paul.kocialkowski@bootlin.com>
Date: Tue, 28 Aug 2018 09:19:08 +0200
Subject: [PATCH 5/7] media: cedrus: h264: Harmonize header

Signed-off-by: Paul Kocialkowski <paul.kocialkowski@bootlin.com>
---
 drivers/staging/media/sunxi/cedrus/cedrus_h264.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/media/sunxi/cedrus/cedrus_h264.c b/drivers/staging/media/sunxi/cedrus/cedrus_h264.c
index 966d20f6599a..e28ffa9c1829 100644
--- a/drivers/staging/media/sunxi/cedrus/cedrus_h264.c
+++ b/drivers/staging/media/sunxi/cedrus/cedrus_h264.c
@@ -1,5 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-or-later
 /*
+ * Cedrus VPU driver
+ *
  * Copyright (c) 2013 Jens Kuske <jenskuske@gmail.com>
  * Copyright (c) 2018 Bootlin
  */
-- 
2.18.1

